biohansel/create/cluster_generator.py

```diff
@@ -0,0 +1,236 @@
+    clustering_array = create_linkage_array(distance_matrix)
+
+    flat_clusters = output_flat_clusters(clustering_array, df.columns, distance_matrix, min_group_size)
+
+    return flat_clusters
+
+
+def compute_distance_matrix(df: pd.DataFrame) -> np.ndarray:
+    """Takes in a binary SNV state DataFrame and outputs a distance matrix using the hamming method
+
+    Args:
+        df: the DataFrame that contains only the samples' binary data
+
+    Returns:
+        a matrix of pair-wise distances between samples
+    """
+
+    return sp.spatial.distance.pdist(
+        df.transpose(), metric='hamming')
+
+
+def create_linkage_array(distance_matrix: np.ndarray) -> np.ndarray:
+    """Takes in a distance matrix and outputs a hierarchical clustering linkage array
+
+    Args:
+        distance_matrix: a matrix of pair-wise distances between samples
+
+    Returns:
+        clustering_array: a hierarchical clustering linkage array that is calculated from the distance matrix
+    """
+
+    clustering_array = linkage(distance_matrix, method='complete')
+
+    return clustering_array
+
+
+def output_flat_clusters(clustering_array: np.ndarray, genomes_only: List, distance_matrix: np.ndarray,
+                         min_group_size: int) -> \
+        pd.DataFrame:
+    """Uses a set of thresholds to output a flat cluster of the linkage array
+
+    Args:
+        clustering_array: a hierarchical clustering linkage array that is calculated from the distance matrix
+        genomes_only: an array of column names for the original SNV DataFrame
+        distance_matrix: a matrix of pair-wise distances between samples
+        min_group_size: the minimum child group size for each new subtype branching point from the parent group
+
+    Returns:
+         flat_clusters: an array of flat clusters from the clustering array
+    """
+    clusters = np.array([
+        fcluster(clustering_array, t=distance, criterion='distance')
+        for distance in np.sort(np.unique(distance_matrix))])
+    cluster_df = pd.DataFrame(np.array(clusters), index=np.sort(np.unique(distance_matrix)))
+    matrix = cluster_df.transpose()
+    matrix.index = genomes_only
+    matrix = matrix.sort_values(
+        by=[cluster_grouping for cluster_grouping in matrix.columns.sort_values(ascending=False)])
+    thresholds = [threshold for threshold in matrix.columns.sort_values(ascending=False)]
+
+    cluster_dict = defaultdict(list)
+
+    for threshold in thresholds:
+        grouping = '-'.join([str(cluster) for cluster in matrix[threshold]])
+        cluster_dict[grouping].append(threshold)
+
+    cluster_matrix = matrix[[thresholds[0] for thresholds in cluster_dict.values()]]
+    cluster_matrix.columns = [index for index, thresholds in enumerate(cluster_matrix.columns)]
+
+    cluster_matrix = cluster_matrix.drop([0], axis=1)
+
+    clusters_genomes_dict = cluster_df_to_dict(cluster_matrix)
+
+    hierarchical_cluster_df = pd.DataFrame(
```

Comment:
This is very dense, please create some local variables. Also it would be useful to return `hierarchical_cluster_df` so users can see how the clusters were assigned at various thresholds. 

Also, try to put the `df` before the rest of the var name to be consistent with the rest of the codebase, thanks.

Reply:
yes this has been addressed in 94b2c0213878ff804e065eb970dc3023030acfcc